src/Database.php

```diff
@@ -408,6 +410,23 @@ public function getWriteConcern()
         return $this->writeConcern;
     }
 
+    /**
+     * Returns the names of all collections in this database
+     *
+     * @see ListCollectionNames::__construct() for supported options
+     * @param array $options
+     * @return Generator
```

Comment:
In the interest of a conservative public API, I wonder if we should just document that we return a Traversable or Iterator. This would be similar to how I only documented a generic CollectionInfoIterator return type for the `listCollections` method. Granted, that was the abstract the user from thinking about legacy and cursor-based responses, but it was also suggesting that the only thing they need care about was that the current element is a CollectionInfo object.

Generator has additional methods like `send()` that I expect would have no use, so the fact that we happen to use a generator here seems like an implementation detail best not revealed.

Reply:
Generators can only be iterated once and can't be rewound, so I think the distinction is important to make sure people don't expect to repeatedly iterate over the same list of names.

Reply:
Ah, good point. That sounds like a possible deviation from the rewindable iterators returned by `listDatabases` and `listCollections`, though. Is it possible to wrap the generator with our CachingIterator. That would save users from having to ever care about this.

Two side notes:

 * https://github.com/mongodb/mongo-php-library/commit/52df7a112e760b595a3485f50c23adf0591b95e8 refactored our CachingIterator to use IteratorIterator instead of Generators, but I believe it forgot to update comments in the file that still refer to `yield`
 * I just found https://github.com/bpolaszek/rewindable-generator, which is a project that addresses this issue with Generators, but it unfortunately uses SPL's CachingIterator, which has its own problems (see [this blog comment](https://blog.calevans.com/2013/12/19/the-php-cachingiterator/#comment-110609), which I found a while ago in [this tweet](https://twitter.com/jmikola/status/923656201610592257)). As such, it actually breaks if you don't fully iterator the Generator before rewinding, which is something our own CachingIterator addresses (an early `rewind()` exhausts the inner iterator).

Reply:
Good point. `ListCollectionNames` now returns the `CallbackIterator` directly since we already have a `CachingIterator` below. `ListDatabaseNames` returns an `ArrayIterator` to iterate over the array of names. This can then always be changed to return a `CachingIterator` or other iterator should the `listDatabases` command ever be changed to returning a cursor.